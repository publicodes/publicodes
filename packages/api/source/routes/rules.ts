import { NewEngine } from '@/types'
import { catchError, PickInObject } from '@/utils'

export function rules(newEngine: NewEngine) {
	const engine = newEngine({})

	const rules = engine.getParsedRules()

	const filteredRules = Object.fromEntries(
		Object.entries(rules).map(([key, val]) => [
			key,
			PickInObject(val, [
				'nodeKind',
				'rawNode',
				'replacements',
				'suggestions',
				'title',
			]),
		])
	)

	return filteredRules
}

export function rulesId(newEngine: NewEngine, id: string) {
	const engine = newEngine({})

	const [error, result] = catchError(() => engine.getRule(id))

	return !error
		? PickInObject(result, ['title', 'suggestions', 'rawNode', 'nodeKind'])
		: { error: { message: error.message } }
}
