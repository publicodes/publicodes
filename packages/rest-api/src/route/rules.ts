import { Engine } from '../types.js'
import { catchError, PickInObject } from '../utils.js'

export function rules(engine: Engine) {
	const rules = engine.getParsedRules()

	const filteredRules = Object.fromEntries(
		Object.entries(rules).map(([key, val]) => [
			key,
			PickInObject(val, [
				'title',
				'nodeKind',
				'rawNode',
				'replacements',
				'suggestions',
			]),
		]),
	)

	return filteredRules
}

type RulesId = string

export function rulesId(engine: Engine, id: RulesId) {
	const [error, result] = catchError(() => engine.getRule(id))

	return !error ?
			PickInObject(result, [
				'title',
				'nodeKind',
				'rawNode',
				'replacements',
				'suggestions',
			])
		:	{ error: { message: error.message } }
}
