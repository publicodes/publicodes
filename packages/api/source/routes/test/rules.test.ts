import Engine from 'publicodes'
import { describe, expect, it } from 'vitest'
import { rules, rulesId } from '../rules'

const newEngine = () => new Engine('rules: 42')

describe('evaluate', () => {
	it('Should list rules', () => {
		expect(rules(newEngine)).toMatchSnapshot()
	})

	it('Should return rules data', () => {
		expect(rulesId(newEngine, 'rules')).toMatchSnapshot()
	})

	it('Should return an error', () => {
		expect(rulesId(newEngine, 'bad rules')).toMatchSnapshot()
	})
})
