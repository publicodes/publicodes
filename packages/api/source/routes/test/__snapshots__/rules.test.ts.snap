// Vitest Snapshot v1

exports[`evaluate > Should list rules 1`] = `
{
  "rules": {
    "nodeKind": "rule",
    "rawNode": {
      "formule": "42",
      "nom": "rules",
    },
    "replacements": [],
    "suggestions": {},
    "title": "Rules",
  },
}
`;

exports[`evaluate > Should return an error 1`] = `
{
  "error": {
    "message": "La rÃ¨gle 'bad rules' n'existe pas",
  },
}
`;

exports[`evaluate > Should return rules data 1`] = `
{
  "nodeKind": "rule",
  "rawNode": {
    "formule": "42",
    "nom": "rules",
  },
  "suggestions": {},
  "title": "Rules",
}
`;
